### PR TITLE
Enable reading ~/.databrickscfg files form MLflow Java API

### PR DESCRIPTION
**Pull request by aarondav**
This PR adds a DatabricksConfigHostCredsProvider which is intended to mimic the profile reading logic from the Databricks CLI:
https://github.com/databricks/databricks-cli/blob/133fea615560c9e4091a785922b127467a1e08ea/databricks_cli/configure/provider.py#L217-L232

This allows the Java SDK to directly connect to Databricks-backed tracking servers given the presence of a `~/.databrickscfg` file.

Additionally, support is added for a barebones "dynamic config provider" which attempts to load a class which, if present, may provide credentials for use. This allows for dynamic refreshable tokens that in Python could be provided by using the `set_config_provider` global API, but which is harder to do in Java due to circular dependencies.

### CODE REVIEW

File: mlflow/java/client/pom.xml [hunk 1]
```diff
@@ -74,6 +82,7 @@
               <include>commons-codec:commons-codec</include>
               <include>commons-logging:commons-logging</include>
               <include>org.apache.httpcomponents:httpcore</include>
+              <include>org.yaml:snakeyaml</include>
```
Comment (mateiz) [inline mlflow/java/client/pom.xml]: Do we actually use YAML? Is it a dependency of commons-io or ini4j?

File: mlflow/java/client/src/main/java/org/ini4j/PyIniConfigParserUtils.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package org.ini4j;
+
+/**
+ * Helper to expose package-private fields of ini4j.
+ */
+public class PyIniConfigParserUtils {
```
Comment (mateiz) [inline mlflow/java/client/src/main/java/org/ini4j/PyIniConfigParserUtils.java]: Do we really need to add this? It seems that ConfigParser has a `get(Strings section, String option)` that we could call directly. Their tutorial actually uses the `Ini` class though, which also lets you get sections via get() (http://ini4j.sourceforge.net/tutorial/IniTutorial.java.html).

File: mlflow/java/client/pom.xml [hunk 3]
```diff
@@ -41,6 +41,14 @@
       <groupId>com.google.protobuf</groupId>
       <artifactId>protobuf-java-util</artifactId>
     </dependency>
+    <dependency>
+      <groupId>commons-io</groupId>
+      <artifactId>commons-io</artifactId>
```
Comment (mateiz) [inline mlflow/java/client/pom.xml]: This dependency seems to only be used in a test, so maybe we can mark it with `<scope>test</scope>`.

File: mlflow/java/client/pom.xml (hunk 1 above)
Comment (aarondav) [reply mlflow/java/client/pom.xml]: Sorry, good catch. Prior implementation.

File: mlflow/java/client/src/main/java/org/ini4j/PyIniConfigParserUtils.java (hunk 2 above)
Comment (aarondav) [reply mlflow/java/client/src/main/java/org/ini4j/PyIniConfigParserUtils.java]: Ah... I was using the `ConfigParser` API, which is apparently a different parser that handles DEFAULTS in an unsuitable way -- that's what I get for googling "ConfigParser ini java". Good catch

File: mlflow/java/client/pom.xml (hunk 3 above)
Comment (aarondav) [reply mlflow/java/client/pom.xml]: Nice!

File: mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    List<String> exceptionMessages = new ArrayList<>();
+    for (MlflowHostCredsProvider provider : hostCredsProviders) {
+      try {
+        MlflowHostCreds hostCreds = provider.getHostCreds();
+
+        if (hostCreds != null && hostCreds.getHost() != null) {
+          logger.debug("Loading credentials from " + provider.toString());
+          return hostCreds;
+        }
+      } catch (Exception e) {
```
Comment (tomasatdatabricks) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java]: Maybe safer to catch Throwable?

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good. I left two minor comments (and I had one inline question) inline.

File: mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.mlflow.tracking.MlflowClientException;
+
+public class HostCredsProviderChain implements MlflowHostCredsProvider {
+  private static final Logger logger = Logger.getLogger(HostCredsProviderChain.class);
+
+  private final List<MlflowHostCredsProvider> hostCredsProviders = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (aarondav) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java]: Hmm, typically Java authors prefer `List` to array for immutable-ish lists. The latter is not immutable, it merely cannot be resized, and its type invariance, lack of a good print function, lack of a good hash function, etc. make it less nice to use in general.

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java [hunk 6]
```diff
@@ -197,8 +196,20 @@ private static String getDefaultTrackingUri() {
...
     MlflowHostCredsProvider provider;
+
     if ("http".equals(uri.getScheme()) || "https".equals(uri.getScheme())) {
       provider = new BasicMlflowHostCreds(trackingUri);
+    } else if (trackingUri.equals("databricks")) {
+      MlflowHostCredsProvider profileProvider = new DatabricksConfigHostCredsProvider();
+      MlflowHostCredsProvider dynamicProvider =
+        DatabricksDynamicHostCredsProvider.createIfAvailable();
+      if (dynamicProvider != null) {
+        provider = new HostCredsProviderChain(dynamicProvider, profileProvider);
```
[the remark this answers is not in the document]
Comment (aarondav) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java]: Yes.

File: mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java (hunk 4 above)
Comment (aarondav) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java]: We don't want to catch things like OutOfMemoryError, as that could cause the JVM to be unstable -- Scala's NonFatal is more suitable, but in Java most authors leave it at Exception.

Comment (tomasatdatabricks) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java]: Yeah OOM makes sense but any Error will not be caught as Exception.

**Review by tomasatdatabricks — APPROVED** [review]
LGTM

File: mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java (hunk 4 above)
Comment (aarondav) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/creds/HostCredsProviderChain.java]: But would be caught by Throwable

File: mlflow/java/client/src/main/java/org/mlflow/tracking/creds/DatabricksConfigHostCredsProvider.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          " using the Databricks CLI.");
+      }
+    }
+    assert (section != null);
+
+    String host = section.get("host");
+    String username = section.get("username");
+    String password = section.get("password");
+    String token = section.get("token");
+    boolean insecure = section.get("insecure", "false").equals("true");
```
Comment (andrewmchen) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/creds/DatabricksConfigHostCredsProvider.java]: Does this need to be converted to lower case before the comparison?

Comment (aarondav) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/creds/DatabricksConfigHostCredsProvider.java]: Let me do this in a follow-up -- good catch!

**Pull request merged**